[ErrorHandler] Rework fatal errors

fancyweb opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/32605
| License       | MIT
| Doc PR        | -

Built on top of https://github.com/symfony/symfony/pull/33038 so review only the second commit : https://github.com/symfony/symfony/pull/33053/commits/d5c3f7ed48e724c88517d616aeabff6ad4255f33

The goals of this PR is to replace current "fatal error handlers" with "error enhancers" since all our current fatal error handlers works on \Error since PHP7.

That means we won't use the FatalErrorException anymore, so we will be able to remove it (once we don't need it in the rest of the codebase).

The final goal btw is to handle \Throwable everywhere in the code so we can remove FatalThrowableError & FatalErrorException classes.

fancyweb committed:
[ErrorHandler] Rework fatal error handlers

fancyweb commented on src/Symfony/Component/ErrorHandler/Error/ClassNotFoundFatalError.php:
```diff
@@ -0,0 +1,37 @@
+
+namespace Symfony\Component\ErrorHandler\Error;
+
+class ClassNotFoundFatalError extends ClassNotFoundError implements FatalErrorInterface
```
This behavior is necessary because of https://3v4l.org/KjaLl (we can get rid of it once min PHP requirement is > 7.3)

stof replied:
You cannot get rid of it, as PHP 7.4 also needs it according to your snippet.

fancyweb replied:
You are right.

fancyweb replied:
It's actually back on 7.4.

Tobion commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -657,6 +664,20 @@ public static function handleFatalError(array $error = null): void
+     *
+     * @return ErrorEnhancerInterface[]
+     */
+    protected function getErrorEnhancers(): array
```
please typehint `iterable` as that is the contract we only need

yceruto commented:
(rebase needed)

fancyweb commented:
Impossible to resolve conflicts, because of 1st commit.

yceruto commented:
@fancyweb Thanks a lot for this PR, great job so far!

There are still some pending failures, let me know when this is ready for you and I will take a deeper look.

stof commented on src/Symfony/Component/ErrorHandler/ErrorEnhancer/ClassNotFoundErrorEnhancer.php:
```diff
@@ -0,0 +1,193 @@
+                }
+            }
+
+            if ($function[0] instanceof ComposerClassLoader || $function[0] instanceof SymfonyClassLoader) {
```
`SymfonyClassLoader` is a dead component in 4.x. Do we still need to account for it in the new ErrorHandler component ?

fancyweb replied:
Talked about this with @nicolas-grekas a while ago, and he told me we can eventually stop supporting it on 5.0. But I didn't do it yet.

yceruto reviewed (approved):
LGTM

fabpot commented:
#33038 has been merged now. @fancyweb Can you rebase this one?

fancyweb commented on src/Symfony/Component/ErrorHandler/Tests/phpt/decorate_exception_handler_lt_74.phpt:
```diff
@@ -0,0 +1,56 @@
+<?php if (PHP_VERSION_ID >= 70400) die('skip: Only for PHP < 7.4.'); ?>
+--INI--
+display_errors=0
+--FILE--
```
Can't extract this in a separate file without altering the composer.json because of the return type patching on 7.4.

fancyweb commented:
@fabpot Rebase is done and CI is green.

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/ErrorEnhancer/ErrorEnhancerInterface.php:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * @return \Error|null An \Error instance if the class is able to improve the error, null otherwise
+     */
+    public function handleError(\Error $error): ?\Error;
```
could it return `FatalErrorInterface`?
(and should we renamed FatalErrorInterface to EnhancedErrorInterface, or just ErrorInterface btw?)

Tobion commented on src/Symfony/Component/ErrorHandler/Error/ClassNotFoundError.php:
```diff
@@ -0,0 +1,33 @@
+            'line' => $previous->getLine(),
+            'trace' => $previous->getTrace(),
+        ] as $property => $value) {
+            $refl = new \ReflectionProperty('Error', $property);
```
```suggestion
            $refl = new \ReflectionProperty(\Error::class, $property);
```

Tobion replied:
Same thing in several other places.

Tobion commented on src/Symfony/Component/ErrorHandler/Error/FatalErrorInterface.php:
```diff
@@ -0,0 +1,20 @@
+interface FatalErrorInterface extends \Throwable
+{
+    /**
+     * @return array An array as returned by error_get_last()
```
No need for the `@return` duplicating the return type.

```suggestion
     * Returns an array as returned by error_get_last()
```

Tobion commented on src/Symfony/Component/ErrorHandler/FatalErrorHandler/FatalErrorHandlerInterface.php:
```diff
@@ -23,9 +23,7 @@ interface FatalErrorHandlerInterface
     /**
      * Attempts to convert an error into an exception.
```
seems wrong now

Tobion commented on src/Symfony/Component/ErrorHandler/FatalErrorHandler/FatalErrorHandlerInterface.php:
```diff
@@ -23,9 +23,7 @@ interface FatalErrorHandlerInterface
+     * @return FatalErrorInterface|null A FatalErrorInterface instance if the class is able to improve the error, null otherwise
      */
-    public function handleError(array $error, FatalErrorException $exception);
+    public function handleError(FatalErrorInterface $fatalError): ?FatalErrorInterface;
```
I think it would be good to get rid of `FatalErrorHandlerInterface`. Can't we handle this with the ErrorEnhancers as well? Maybe that requires to change the type hints in ErrorEnhancers from \Error to \Throwable. But that seems fine anyway because the ErrorEnhancers could also be used for Exceptions in theory.

fancyweb replied:
Yes!

Tobion commented on src/Symfony/Component/ErrorHandler/ErrorEnhancer/ErrorEnhancerInterface.php:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * @return \Error|null An \Error instance if the class is able to improve the error, null otherwise
+     */
+    public function handleError(\Error $error): ?\Error;
```
`handleError` is not descriptive and confusing IMO. How about `enhanceError`?

fancyweb replied:
Just `enhance` ?

fancyweb commented:
Just repushed to take comments into account:

1. Removed `FatalErrorHandlerInterface`, everything is done by `ErrorEnhancerInterface`.
2. Removed `FatalErrorInterface` and everything related that was only necessary to know if a class not found error was "fatal" or not. We handle it but end user just can't know if the error he gets was fatal or not (well not easily).
3. Removed `FatalErrorException` and `FatalThrowableError` in favor of `ErrorException`.

Tobion commented:
Thanks @fancyweb for working on this feature, this is much appreciated.

yceruto commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -13,15 +13,13 @@
-use Symfony\Component\ErrorHandler\FatalErrorHandler\UndefinedFunctionFatalErrorHandler;
-use Symfony\Component\ErrorHandler\FatalErrorHandler\UndefinedMethodFatalErrorHandler;
-use Symfony\Component\ErrorRenderer\ErrorRenderer\HtmlErrorRenderer;
-use Symfony\Component\ErrorRenderer\Exception\FlattenException;
```
Unintentionally deleted, fixed in https://github.com/symfony/symfony/pull/34075
